File: Level_4(C++,Java)/java/Static/Passport/Passport.java
public class Passport {
    private int id;
    private String series;
    private String name;
    private String surname;
    private Date dateOfBirth;
    
    private static final int startId = 100000;
    private static final int finalId = 999999;
    private static int lastId = startId;
    private static char[] lastSeries = new char[]{'A','A'};
    
    private void validateSeries() throws OutOfSeriesException {
        if ( lastSeries[0] == 'Z' && lastSeries[1] == 'Z' && lastId > finalId ) {
            throw new OutOfSeriesException();
        }
        if ( lastSeries[1] == 'Z' && lastId > finalId ) {
            lastSeries[0] += 1;
            lastSeries[1] = 'A';
            lastId = startId;
        } else if ( lastId > finalId ) {
            lastSeries[1] += 1;
            lastId = startId;
        }
    }

    private static void validateNewSeries(String newSeries) throws InvalidSeriesException {
        if ( newSeries.length() != lastSeries.length ) {
            throw new InvalidSeriesException();
        }
        if ( newSeries.charAt(0) < lastSeries[0] || newSeries.charAt(0) > 'Z' ) {
            throw new InvalidSeriesException();
        }
        if ( newSeries.charAt(1) < 'A' || newSeries.charAt(1) > 'Z' ) {
            throw new InvalidSeriesException();
        }
        if ( newSeries.charAt(0) == lastSeries[0] && newSeries.charAt(1) < lastSeries[1] ) {
            throw new InvalidSeriesException();
        }
    }
    
    public Passport(String surname, String name, Date dateOfBirth) throws OutOfSeriesException {
        this.validateSeries();
        this.id = lastId;
        this.series = new String(lastSeries);
        this.name = name;
        this.surname = surname;
        this.dateOfBirth = dateOfBirth;
        lastId += 1;
    }

    public Passport(Passport copy) throws OutOfSeriesException {
        this.validateSeries();
        this.id = lastId;
        this.series = new String(lastSeries);
        this.name = copy.name;
        this.surname = copy.surname;
        this.dateOfBirth = copy.dateOfBirth;
        lastId += 1;
    }
    
    public int getId() {
        return this.id;
    }

    public String getSeries() {
        return this.series;
    }

    public String getName() {
        return this.name;
    }

    public String getSurname() {
        return this.surname;
    }

    public Date getDate() {
        return this.dateOfBirth;
    }
    
    public static void setSeries(String newSeries) throws InvalidSeriesException {
        validateNewSeries(newSeries);
        
        lastSeries[0] = newSeries.charAt(0);
        lastSeries[1] = newSeries.charAt(1);
        lastId = startId;
    }

    public static void setSeries(String newSeries, int newId) throws InvalidSeriesException, InvalidIdException {
        validateNewSeries(newSeries);
        
        if ( newId < startId || newId > finalId ) {
            throw new InvalidIdException();
        }
        
        lastSeries[0] = newSeries.charAt(0);
        lastSeries[1] = newSeries.charAt(1);
        lastId = newId;
    }
    
    public String toString() {
        StringBuffer out = new StringBuffer();
        
        out.append("--------------------------------\n");
        out.append("     Passport: ");
        out.append(this.series);
        out.append(" ");
        out.append(this.id);
        out.append("\n");
        out.append("      Surname: ");
        out.append(this.surname);
        out.append("\n");
        out.append("         Name: ");
        out.append(this.name);
        out.append("\n");
        out.append("Date of birth: ");
        out.append(this.dateOfBirth);
        
        return out.toString();
    }
}

File: Level_4(C++,Java)/C++/static/Passport/Passport.cpp
#include "Passport.h"

const int Passport::startId = 100000;
const int Passport::finalId = 999999;
int Passport::lastId = startId;
std::string Passport::lastSeries = "AA";

void Passport::validateSeries() {
    if ( lastSeries[0] == 'Z' && lastSeries[1] == 'Z' && lastId > finalId ) {
        throw OutOfSeries();
    }
    if ( lastSeries[1] == 'Z' && lastId > finalId ) {
        lastSeries[0] += 1;
        lastSeries[1] = 'A';
        lastId = startId;
    } else if ( lastId > finalId ) {
        lastSeries[1] += 1;
        lastId = startId;
    }
}

void Passport::validateNewSeries(const std::string& newSeries) {
    if ( newSeries.size() != lastSeries.size() ) {
        throw InvalidSeries();
    }
    if ( newSeries[0] < lastSeries[0] || newSeries[0] > 'Z' ) {
        throw InvalidSeries();
    }
    if ( newSeries[1] < 'A' || newSeries[1] > 'Z' ) {
        throw InvalidSeries();
    }
    if ( newSeries[0] == lastSeries[0] && newSeries[1] < lastSeries[1] ) {
        throw InvalidSeries();
    }
}

Passport::Passport(const std::string& surname, const std::string& name, const Date& dateOfBirth) {
    validateSeries();
    this->id = lastId;
    this->series = lastSeries;
    this->name = name;
    this->surname = surname;
    this->dateOfBirth = dateOfBirth;
    lastId += 1;
}

Passport::Passport(const Passport& copy) {
    validateSeries();
    this->id = lastId;
    this->series = lastSeries;
    this->name = copy.name;
    this->surname = copy.surname;
    this->dateOfBirth = copy.dateOfBirth;
    lastId += 1;
}

Passport::~Passport() {}

int Passport::getId() const {
    return this->id;
}

const std::string& Passport::getSeries() const {
    return this->series;
}

const std::string& Passport::getName() const {
    return this->name;
}

const std::string& Passport::getSurname() const {
    return this->surname;
}

const Date& Passport::getDate() const {
    return this->dateOfBirth;
}

Passport& Passport::operator=(const Passport& copy) {
    this->name = copy.name;
    this->surname = copy.surname;
    this->dateOfBirth = copy.dateOfBirth;
    
    return *this;
}

void Passport::setSeries(const std::string& newSeries) {
    validateNewSeries(newSeries);
    
    lastSeries = newSeries;
    lastId = startId;
}

void Passport::setSeries(const std::string& newSeries, int newId) {
    validateNewSeries(newSeries);
    
    if ( newId < startId || newId > finalId ) {
        throw InvalidId();
    }
    
    lastSeries = newSeries;
    lastId = newId;
}

std::ostream& operator<<(std::ostream& out, const Passport& pass) {
    out << "--------------------------------" << '\n';
    out << "     Passport: " << pass.getSeries() << ' ' << pass.getId() << '\n';
    out << "      Surname: " << pass.getSurname() << '\n';
    out << "         Name: " << pass.getName() << '\n';
    out << "Date of birth: " << pass.getDate();
}
